engine-comparison/begin-experiment.sh

```diff
@@ -5,59 +5,69 @@
-else
-  BENCHMARKS=$(echo $1 | tr ',' ' ')
-fi
-
-echo "Restarting gcloud instance. Instance should already be created (with gcloud_creator.sh)"
 # Start one gcloud instance for the dispatcher
-gcloud compute instances start $INSTANCE_NAME --zone=$GCLOUD_ZONE
-
-# Can't ssh immediately
-while :
-do
-  sleep 6 # arbitrary
-  ! (gcloud compute ssh $INSTANCE_NAME --command="mkdir ~/input" --zone=$GCLOUD_ZONE \
-    2>&1 | grep "ERROR") && break # Break loop as soon as there is no error
-  echo "GCloud VM isn't ready yet. Rerunning SSH momentarily"
-done
+# echo "Restarting gcloud instance. Instance should already be created (with gcloud_creator.sh)"
+create_or_start $INSTANCE_NAME
+robust_begin_gcloud_ssh $INSTANCE_NAME
+gcloud compute ssh $INSTANCE_NAME --command="mkdir ~/input" --zone=$GCLOUD_ZONE
 
 # Send configs for the fuzzing engine
 FENGINE_CONFIGS=${@:2}
 
-if [[ -d tmp-configs ]]; then
-  rm -r tmp-configs
+if [[ -d fengine-configs ]]; then
+  rm -r fengine-configs
 fi
-mkdir tmp-configs
+mkdir fengine-configs
 for FENGINE in $FENGINE_CONFIGS; do
-  cp $FENGINE tmp-configs/$FENGINE
+  cp $FENGINE fengine-configs/$FENGINE
 done
-gcloud compute scp tmp-configs/ ${INSTANCE_NAME}:~/input --recurse --zone=$GCLOUD_ZONE
-rm -r tmp-configs
+
+gcloud compute scp fengine-configs/ ${INSTANCE_NAME}:~/input --recurse --zone=$GCLOUD_ZONE
+rm -r fengine-configs
+
+# These will frequently/usually be defined by the user
+JOBS=${JOBS:-8}
+N_ITERATIONS=${N_ITERATIONS:-5}
+
+# Send configs
+AUTOGEN=${SCRIPT_DIR}/autogen
+
+if [[ ! -d $AUTOGEN ]]; then
+  mkdir $AUTOGEN
+fi
+
+if [[ ! -e ${AUTOGEN}/dispatcher.config ]]; then
+  echo "BMARKS=$1" > ${AUTOGEN}/dispatcher.config
+fi
+
+if [[ ! -e ${AUTOGEN}/worker.config ]]; then
+  echo "N_ITERATIONS=$N_ITERATIONS" > ${AUTOGEN}/worker.config
+  echo "JOBS=$JOBS" >> ${AUTOGEN}/worker.config
+fi
+# Pass service account auth key
+if [[ ! -e ${AUTOGEN}/dispatcher-key.json ]]; then
+  gcloud iam service-accounts keys create ${AUTOGEN}/dispatcher-key.json \
+    --iam-account=$SERVICE_ACCOUNT --key-file-type=json
+fi
 
 # Send the entire local FTS repository to the dispatcher;
 # Local changes to any file will propagate
 gcloud compute scp $(dirname $SCRIPT_DIR) ${INSTANCE_NAME}:~/input --recurse --zone=$GCLOUD_ZONE
-gcloud compute ssh $INSTANCE_NAME --command="mv ~/input/$(basename $(dirname ${SCRIPT_DIR})) ~/input/FTS" --zone="$GCLOUD_ZONE"
+
+# Could use "! [[ -d ~/input/FTS ]] &&" to prevent deletion
+gcloud compute ssh $INSTANCE_NAME --command="rm -rf ~/input/FTS && \
+  mv ~/input/$(basename $(dirname ${SCRIPT_DIR})) ~/input/FTS " --zone=$GCLOUD_ZONE
 
 # Run dispatcher with Docker
-DISPATCHER_COMMAND="sudo docker build -f ~/input/FTS/engine_comparison/ --build-arg run-script=dispatcher.sh ~/input"
+DISPATCHER_COMMAND="~/input/FTS/engine-comparison/run.sh /~/work/FTS/engine-comparison/dispatcher.sh"
 gcloud compute ssh $INSTANCE_NAME --command="$DISPATCHER_COMMAND" --zone=$GCLOUD_ZONE
 
```

Comment:
You're right that it's not always necessary. I've realized that at one point I had issues because I skipped the part of `gcloud init` where I set my zone. In general, it may be a safe assumption to make, but setting the zone is strictly optional in gcloud init. Because you mention this, I've realized that I should just create a robust method for solving this